RFC-0004-support-distributed-procedure.md

```diff
@@ -0,0 +1,397 @@
+
+```java
+public interface Metadata
+{
+    ......
+
+    /**
+     * Begin call distributed procedure
+     */
+    DistributedProcedureHandle beginCallDistributedProcedure(Session session, QualifiedObjectName procedureName, TableHandle tableHandle, Object[] arguments);
+
+    /**
+     * Finish call distributed procedure
+     */
+    void finishCallDistributedProcedure(Session session, DistributedProcedureHandle procedureHandle, QualifiedObjectName procedureName, Collection<Slice> fragments);
+
+    ......
+}
+
+public interface ConnectorMetadata
+{
+    ......
+
+    /**
+     * Begin call distributed procedure
+     */
+    default ConnectorDistributedProcedureHandle beginCallDistributedProcedure(
+            ConnectorSession session,
+            QualifiedObjectName procedureName,
+            ConnectorTableLayoutHandle tableLayoutHandle,
+            Object[] arguments)
+    {
+        throw new PrestoException(NOT_SUPPORTED, "This connector does not support distributed procedure");
+    }
+
+    /**
+     * Finish call distributed procedure
+     *
+     * @param fragments all fragments returned by {@link com.facebook.presto.spi.UpdatablePageSource#finish()}
+     */
+    default void finishCallDistributedProcedure(ConnectorSession session, ConnectorDistributedProcedureHandle procedureHandle, QualifiedObjectName procedureName, Collection<Slice> fragments)
+    {
+        throw new PrestoException(NOT_SUPPORTED, "This connector does not support distributed procedure");
+    }
+
+    ......
+}
+```
+
+Wherein, `beginCallDistributedProcedure` is used for preparation work before the start of distributed scheduling and generate the execution information that needs to be passed to worker nodes. And `finishCallDistributedProcedure` is used for the whole transaction submission after the completion of distributed writing.
+
+As for a specific connector (such as Iceberg), in the implementation logic of method `beginCallDistributedProcedure` and `finishCallDistributedProcedure` in `ConnectorMetadata`, in addition to accomplish the common logic for this connector (such as starting a transaction, building a transaction context, committing a transaction, etc.), it should also resolve the specified distributed procedure and call its relevant method to execute the procedure-customized logic.
+
+#### 6. Iceberg connector's support for distributed procedure
+
+In Iceberg, we often need to record the original data files that have been scanned and rewritten during the execution of table data operations (including deleted files that have been fully applied), and in the final submission, combine the newly generated data files due to rewriting to make some changes and transaction submissions at the metadata level.
```

Comment:
Iceberg in Native engine uses HiveConnector itself. The HiveConnector in Native engine handles both Hive and Iceberg splits. Does HiveConnector support distributed procedure ? How will it be reused or enhanced for Iceberg ?

Reply:
As `beginCallDistributedProcedure` and `finishCallDistributedProcedure` in `ConnectorMetadata` are all invoked in coordinator, I think the main change for hive to support distributed procedures is in java implementation.

Hive should implement `beginCallDistributedProcedure` and `finishCallDistributedProcedure` in `ConnectorMetadata`, customize it's own preparation and submission logic, and then implement and register its own distributed procedures. All these jobs could be done in java only.

So I think the main change in worker end logic which need to support native c++ is that it should generate and provide `ConnectorPageSink` based on the newly added `ConnectorDistributedProcedureHandle`. Please let me know if there are any omissions.